### PR TITLE
Item meta pointing range

### LINKED ISSUE #10164: Add per-ItemStack point range
## Problem

This is a problem. As a modder, I and probably many others as well, have wanted a way to change the range (distance for digging and placing nodes, and punching) of the hand tool at runtime. However, it can only be set by overriding the hand tool while loading a mod. I know this isnt something a whole lot of nodders may take advantage of, but it could definitely help with custom games.

## Solutions

For stacks and stack meta refs, you can use

ItemStack:get_toolcapabilities()
 and
ItemStackMetaRef:set_toolcapabilities(tool_caps)

IT would be nice to also have

ItemStack:get_range()
 and
ItemStack:set_range(range) --or nil to reset it to default

## Additional context

The point of this is also to be able to change the range of an ItemStack at runtime.
I know that you could just create a tool that has a farther range but this is very different and does not solve the problem. Please consider this. Thanks.

### PR DESCRIPTION
**Pull request by cx384**
## Goal of this PR
- Resolves #10164
- The author of this issue already made a PR, but it got closed due to inactivity, I think. #13566
  (The PR got a Concept approved label, but the issue not.)
- So this PR is a new try to solve it.

## How does it work
It uses the item meta key `range` to determine the pointing range.

## To do

This PR Ready for Review.

## How to test

Start devtest, and left click with the `testitems:telescope_stick` to try out different ranges.

### CODE REVIEW

**Comment by Zughy** [comment]
@cx384 rebase needed

**Comment by cx384** [comment]
Rebase done.

File: doc/lua_api.md [hunk 1]
```diff
@@ -2540,6 +2540,7 @@ Some of the values in the key-value store are handled specially:
   0 = default, 1 = left / up, 2 = middle, 3 = right / down
   The default currently is the same as right/down.
   Example: 6 = 2 + 1*4 = middle,up
+* `range`: Overrides the pointing range
```
Comment (appgurueu) [inline doc/lua_api.md:2543]: This is a bit brief. Maybe give an example, say that it's a float?

File: games/devtest/mods/testitems/init.lua [hunk 2]
```diff
@@ -90,3 +90,18 @@ minetest.register_craftitem("testitems:image_meta", {
...
+minetest.register_craftitem("testitems:telescope_stick", {
+	description = S("Telescope Stick (Increases range on use.)"),
+	inventory_image = "testitems_telescope_stick.png",
+	on_use = function(itemstack, player)
+		local meta = itemstack:get_meta()
+		local range = (tonumber(meta:get_string("range")) or 0) + 1.2
+		if range > 10 then
+			range = 0
+		end
+		meta:set_string("range", range)
```
Comment (appgurueu) [inline games/devtest/mods/testitems/init.lua:103]: I think it would be slightly cleaner to use `set_float` here.

File: games/devtest/mods/testitems/init.lua [hunk 3]
```diff
@@ -90,3 +90,18 @@ minetest.register_craftitem("testitems:image_meta", {
 		return itemstack
 	end,
 })
+
+minetest.register_craftitem("testitems:telescope_stick", {
+	description = S("Telescope Stick (Increases range on use.)"),
+	inventory_image = "testitems_telescope_stick.png",
+	on_use = function(itemstack, player)
+		local meta = itemstack:get_meta()
+		local range = (tonumber(meta:get_string("range")) or 0) + 1.2
```
Comment (appgurueu) [inline games/devtest/mods/testitems/init.lua:99]: You can use `get_float` instead of `tonumber(... or 0)` here.

**Review by appgurueu — COMMENTED** [review]
I agree that it would most probably be cleaner to pass the item def manager. That would allow us to get rid of one parameter. The call site would also be more readable: `getToolRange(selected_item.getDefinition(m_itemdef), hand_item.getDefinition(m_itemdef), selected_item, hand_item)` would simplify to `getToolRange(m_itemdef, hand_item, selected_item)`.

The telescope stick works as expected.

**Comment by cx384** [comment]
I applied your suggestions.

> Regarding your note, I think that it would most probably be cleaner to pass the item def manager. There is no performance concern here.

Now it gets an item def manager and `processPlayerInteraction` calls `selected_item.getDefinition` multiple times. If I'm not wrong, it is just a lookup in a red black tree. So as you said, no performance concerns.

All fields stored in MetaDataRef are just strings, so it doesn't matter how they are set, but I changed it for clarification.

**Review by appgurueu — APPROVED** [review]

**Comment by appgurueu** [comment]
> All fields stored in MetaDataRef are just strings, so it doesn't matter how they are set, but I changed it for clarification.

Indeed, in practice this isn't an issue. FYI there are some ugly details though (for example, your `set_string` call was relying on number -> string coercion, which in turn was relying on `tostring` to properly format a float, which in practice only works because we force the C locale for number formatting) so I think using `set_float` / `get_float` and letting the engine handle the formatting is the cleanest option here.

File: doc/lua_api.md [hunk 4]
```diff
@@ -5316,6 +5318,8 @@ Utilities
       dynamic_add_media_startup = true,
       -- dynamic_add_media supports `filename` and `filedata` parameters (5.9.0)
       dynamic_add_media_filepath = true,
+      -- Overrideable pointing range using the itemstack meta key `"range"`
```
Comment (sfan5) [inline doc/lua_api.md:5321]:
```suggestion
      -- Overrideable pointing range using the itemstack meta key `"range"` (5.9.0)
```

File: src/client/game.cpp [hunk 5]
```diff
@@ -1461,7 +1461,7 @@ void Game::copyServerClientCache()
 {
 	// It would be possible to let the client directly read the media files
 	// from where the server knows they are. But aside from being more complicated
-	// it would also *not* fill the media cache and cause slower joining of 
+	// it would also *not* fill the media cache and cause slower joining of
```
Comment (sfan5) [inline src/client/game.cpp:1464]: it would be good to configure your editor to not whitespace-fix unrelated lines

File: src/tool.cpp [hunk 6]
```diff
@@ -29,7 +29,8 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/hex.h"
 #include "common/c_content.h"
 #include <json/json.h>
-	
+#include <string>
```
Comment (sfan5) [inline src/tool.cpp:32]: tool.h already includes this

**Review by sfan5 — COMMENTED** [review]
lgtm otherwise

File: src/tool.cpp (hunk 6 above)
Comment (cx384) [reply src/tool.cpp]: Removed.

File: doc/lua_api.md (hunk 4 above)
Comment (cx384) [reply doc/lua_api.md]: Fixed.

File: src/client/game.cpp (hunk 5 above)
Comment (cx384) [reply src/client/game.cpp]: I don't think my editor has this feature, maybe I should use git hooks for this.

I readded the whitespace, though I don't think it would hurt to remove it.

**Comment by cx384** [comment]
Thanks for the review, I fixed everything and rebased.
